clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -972,16 +975,28 @@ private Record nextFetchedRecord() {
         private List<ConsumerRecord<K, V>> fetchRecords(int maxRecords) {
             if (isFetched)
                 return Collections.emptyList();
+            if (nextInlineException != null) {
```

Comment:
Looking at the current code, and it seems a little dangerous now to skip a bad message / bad batch now. If the skipped message happens to be a transaction abort control message and the consumer is in a read committed only mode, we may skip the later messages are actually not aborted. It seems the safest option here is to not skip a bad message even after an exception is thrown.

That said, skipping the bad message after throwing exception has long been an existing behavior. @lindong28 had a point that this patch is to make sure the messages before a bad message can be still delivered correctly, which is orthogonal to whether to skip a bad message or not. So maybe we can get this patch in first and then fix the skipping bad messages behavior.

@hachikuji @ijuma Thought?